iceoryx_hoofs/source/posix_wrapper/periodic_timer.cpp

```diff
@@ -0,0 +1,188 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/posix_wrapper/periodic_timer.hpp"
+#include "iceoryx_hoofs/posix_wrapper/posix_call.hpp"
+
+using namespace iox::units::duration_literals;
+
+namespace iox
+{
+namespace posix
+{
+PeriodicTimer::PeriodicTimer(const iox::units::Duration interval) noexcept
+    : m_interval(interval)
+{
+    auto timerSemaphore = posix::Semaphore::create(posix::CreateUnnamedSharedMemorySemaphore, 0U);
+    cxx::Ensures(!timerSemaphore.has_error() && "Could not create Semaphore for PeriodicTimer!");
+    m_waitSemaphore = std::move(timerSemaphore.value());
+    start();
+}
+
+void PeriodicTimer::start() noexcept
+{
+    stop();
+    cxx::Ensures(!m_waitSemaphore.timedWait(m_interval).has_error());
+    auto currentTime = now();
+    cxx::Ensures(!currentTime.has_error());
+    m_timeForNextActivation = currentTime.value() + m_interval;
+}
+
+void PeriodicTimer::start(const iox::units::Duration interval) noexcept
+{
+    m_interval = interval;
+    start();
+}
+
+void PeriodicTimer::stop() noexcept
+{
+    auto semValue = m_waitSemaphore.getValue();
+    cxx::Ensures(!semValue.has_error());
+    if (semValue.value() == iox::posix::SEM_ACQUIRED)
+    {
+        cxx::Ensures(!m_waitSemaphore.post().has_error());
+    }
+}
+
+cxx::expected<units::Duration, TimerErrorCause> PeriodicTimer::now() noexcept
+{
+    struct timespec ts;
+    auto result =
+        posixCall(clock_gettime)(CLOCK_REALTIME, &ts).failureReturnValue(-1).evaluate();
+
+    if (result.has_error())
+    {
+        return createErrorCodeFromErrNo(result.get_error().errnum);
+    }
+
+    return cxx::success<units::Duration>(ts);
+}
+
+cxx::expected<iox::posix::WaitResult, TimerErrorCause> PeriodicTimer::wait(TimerCatchupPolicy policy) noexcept
+{
+    // To check if the TIMER is active (if the sempahore is acquired)
+    auto semValue = m_waitSemaphore.getValue();
+    cxx::Ensures(!semValue.has_error());
+    if (semValue.value() == iox::posix::SEM_ACQUIRED)
+    {
+        auto currentTime = now();
+        cxx::Ensures(!currentTime.has_error());
+        if (currentTime.value() > m_timeForNextActivation)
+        {
+            switch (policy)
+            {
+            case iox::posix::TimerCatchupPolicy::IMMEDIATE_TICK:
+            {
+                m_timeForNextActivation = currentTime.value();
+                m_waitResult.state = iox::posix::TimerState::TICK;
+                return cxx::success<iox::posix::WaitResult>(m_waitResult);
+            }
+
+            case iox::posix::TimerCatchupPolicy::SKIP_TO_NEXT_TICK:
+            {
+                auto delay = currentTime.value() - m_timeForNextActivation; // calculate the time delay
+                if (delay > m_interval)
+                {
+                    auto totalSlotToBeSkipped =
+                        delay.toMilliseconds() / m_interval.toMilliseconds(); // calculate the total slots missed
```

Comment:
I think it must be `auto totalSlotToBeSkipped = delay.toMilliseconds() / m_interval.toMilliseconds() + 1;`

Let's say delay is 1.5 and interval is 1. Then `totalSlotToBeSkipped == 1` and `m_timeForNextActivation += m_interval * 1` which would be a timepoint in the past.
When you then call `m_waitSemaphore.timedWait` you would immediately return and would have always `IMMEDIATE_TICK` behavior.

Maybe I misunderstood this here, but if not please fix it and add a tests. The best way of fixing this would be a test which shows that this is `IMMEDIATE_TICK` behavior and is red until this is fixed.